### DISCUSSION TITLE
How to extend User model with a Field that is ForeignKey

### DISCUSSION THREAD

**Discussion by Tur8008** [fastapi-users/fastapi-users#1160] [Q&A]
Hello,
I need to extend User model with a ForeignKey field and then override User Registration logic depending on that Foreign Key's state. What is the best solution in this case?

In fact, I need to confirm a user's phone before they will be registered and if the number exists in DB is confirmed the registration is allowed. The phone number is a dedicated table which has 2 attrs - phone_number and is_confirmed field.

**Comment by frankie567** [comment]
Well, that's a quite specific case you have here.

Intuitively, I would say it would be easier to do the contrary: the user registers, but its number is not verified. Until its number is not verified, they couldn't do anything on the application. This is what we actually do with the verify email mechanism.

If you really don't want the user to be created before its number is verified, you'll need several things:

* A "pre-registration" endpoint accepting a phone number which:
    * Save the number in database
    * Send the SMS validation code
    * Generate and return to the user a sort of "session state"
* The registration endpoint should be implemented so it requires this "session state" in the payload
    * Thanks to this "session state", you are able to link it to the phone number created in the previous step
    * Check if this phone number has been validated
    * If it is, create the user and link the phone number to it

So with this solution you'll probably need to:

* Implement your own flavor of `/register` route: https://github.com/fastapi-users/fastapi-users/blob/master/fastapi_users/router/register.py
* Overload the `create` method of `UserManager`: https://github.com/fastapi-users/fastapi-users/blob/024e6cd4516a55c2205fc58f719db1e581dd8078/fastapi_users/manager.py#L110-L147

**Comment by Tur8008** [comment]
Thank's for such a detailed answer! I'll follow your advice.

**Comment by Tur8008** [comment]
But even if I trigger on_after_register I still need to extend User model with ForeinKey that can't be bound automatically on post request. So in both scenarios I need to override used_register router so that it fill ForeinKey field in User model. Am I right?

**Reply by frankie567** [reply]
Not necessarily: the `create` method of `UserManager` receives a Pydantic schema, `UserCreate`. It can be any data structure you choose. Then, inside the `create` method, nothing stops you from taking that data and split it into several SQLAlchemy objects.

Schematically, your `UserCreate` could look like this:

```py
class UserCreate(schemas.BaseUserCreate):
    phone_number: str
```

And inside `create`:

```py
phone_number = PhoneNumber(phone_number=user_create.phone_number)
created_user = await self.user_db.create(user_dict)
created_user.phone_number = phone_number
```

**Reply by Tur8008** [reply]
Ok, I'll tackle this. Thank you!

**Reply by Tur8008** [reply]
May be it would be reasonable to add a method like on_before_register to perform some pre-actions before creation?

**Comment by Tur8008** [comment]
I managed to implement writing phone into db as you had suggested before.

```python
  phone_number: str = user_dict.pop("phone_number")
  phone = Phone(phone_number=phone_number, is_confirmed=False)
  user_dict["phone"] = phone
  created_user = await self.user_db.create(user_dict)
  ```
 
 But in fact I need to get phone instance fron db by phone_number. So I need to get db_session somehow. Something like that:
 
```python
   phone_number: str = user_dict.pop("phone_number")
   phone = await self.phone_db.get_by_phone_number(phone_number)
   if not phone or not phone.is_confirmed:
       raise SomeException ...
   user_dict["phone"] = phone
   created_user = await self.user_db.create(user_dict)
  ```
 
  I wonder if I should override get_user_db to add Phone table support or there is a better aproach?

**Reply by frankie567** [reply]
You can very well customize how you instantiate your `UserManager` and pass it a raw SQLAlchemy session:

```py
class UserManager(UUIDIDMixin, BaseUserManager[User, uuid.UUID]):
    def __init__(self, user_db, session):
        self.session = session
        super().__init__(user_db)


async def get_user_manager(user_db=Depends(get_user_db), session: AsyncSession = Depends(get_async_session)):
    yield UserManager(user_db, session)
```

**Reply by Tur8008** [reply]
Looks brilliantly! :)

**Comment by Tur8008** [comment]
Hello,  I' ve just got a related problem. 
I add FK field called Phone and I need this field to be included in response on /users/me and /auth/register endpoints.
Unfortunately I get 
``` json
{
    "id": "ea764ef1-430f-48b0-b10a-c18e2a91a97c",
    "email": "art8008@gmail.com",
    "is_active": true,
    "is_superuser": false,
    "is_verified": false,
    "phone_number": null
}
```
I guess the problem is in router
```python
@router.get(
        "/me",
        response_model=user_schema,
        name="users:current_user",
        responses={
            status.HTTP_401_UNAUTHORIZED: {
                "description": "Missing token or inactive user.",
            },
        },
    )
    async def me(
        user: models.UP = Depends(get_current_active_user),
    ):
        return user_schema.from_orm(user)
```
I can adjust user_chema so it included phone_namber but I have no idea how to fill the schema from_orm when phone is a FK to a db table but not just a string.

For example, what I've managed to do with register endpoint is 
```python
async def register(
        request: Request,
        user_create: user_create_schema,  # type: ignore
        user_manager: BaseUserManager[models.UP, models.ID] = Depends(get_user_manager),
    ):
        try:
            created_user = await user_manager.create(user_create, safe=True, request=request)
        except exceptions.UserAlreadyExists:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail=ErrorCode.REGISTER_USER_ALREADY_EXISTS,
            )
        except exceptions.InvalidPasswordException as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail={
                    "code": ErrorCode.REGISTER_INVALID_PASSWORD,
                    "reason": e.reason,
                },
            )
        # My changes is here, below
        except PhoneNotConfirmed:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail={"code": ExtErrorCode.PHONE_NOT_CONFIRMED},
            )
        return_value = user_schema.from_orm(created_user). #  I changed an original code here <==
        return_value.phone_number = created_user.phone_id
        return return_value
```
I can't get my head around how to implement my changes in a better way.

**Reply by frankie567** [reply]
In an SQLAlchemy model, you need to add a `relationship` attribute so SQLAchelmy automatically queries the related table and add it as a property on `User` model. Something like:

```py
class PhoneNumber(Base):
    phone_number = mapped_column(String(255))
    is_verified = mapped_column(Boolean())

class User(SQLAlchemyBaseUserTableUUID, Base):
    # ...
    phone_number_id = mapped_column(ForeignKey("phone_number.id"))
    phone_number = relationship("PhoneNumber", lazy="joined")
```

`lazy=joined` is important to make SQLAlchemy eagerly load the relationship upfront.

In this approach, `phone_number` will be a `PhoneNumber` object, will all its attributes. So an approach could be to have a schema like this:

```py
class PhoneNumber(BaseModel):
    phone_number: str
    is_verified: bool

class UserRead(schemas.BaseUser[uuid.UUID]):
    # ...
    phone_number: PhoneNumber
```

And the result in JSON would be:

```json
{
    "id": "ea764ef1-430f-48b0-b10a-c18e2a91a97c",
    "email": "art8008@gmail.com",
    "is_active": true,
    "is_superuser": false,
    "is_verified": false,
    "phone_number": {
        "phone_number": "1234567890",
        "is_verified": true
    }
}
```

If you really want to have directly the phone number value, you could use a `hybrid_property`:

```py
class User(SQLAlchemyBaseUserTableUUID, Base):
    # ...
    phone_number_id = mapped_column(ForeignKey("phone_number.id"))
    phone_number_relationship = relationship("PhoneNumber", lazy="joined")

    @hybrid_property
    def phone_number(self):
        return self.phone_number_relationship.phone_number
```
 
Ref:

* https://docs.sqlalchemy.org/en/20/orm/quickstart.html#declare-models
* https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html
* https://docs.sqlalchemy.org/en/20/orm/extensions/hybrid.html#defining-expression-behavior-distinct-from-attribute-behavior

**Reply by Tur8008** [reply]
Yeah! It works! Thank you! I didn't even have to change router's code. Nice!

**Reply by joachimhodana** [reply]
Hey, I tried extending my schema with this approach but it's giving me following error: `Input should be a valid dictionary or instance of PreferencesRead [type=model_type, input_value=<parkr.db.models.models.P...t at 0x00000243F021A7D0>, input_type=Preferences]`. Did i miss something?

My models:
```py
class User(SQLAlchemyBaseUserTableUUID, Base):
    display_name: Mapped[str] = mapped_column(String(255), nullable=False)
    avatar_url: Mapped[str] = mapped_column(String(), nullable=True)
    birthday: Mapped[Date] = mapped_column(Date(), nullable=True)

    oauth_accounts: Mapped[List[OAuthAccount]] = relationship(
        "OAuthAccount",
        lazy="joined",
    )
    preferences: Mapped["Preferences"] = relationship(
        "Preferences",
        back_populates="user",
        lazy="joined",
    )
    cars: Mapped[List["Car"]] = relationship("Car", back_populates="user")


class Preferences(Base):
    __tablename__ = "preferences"
    user_id: Mapped[uuid.UUID] = mapped_column(
        ForeignKey("user.id"),
        nullable=False,
        primary_key=True,
    )
    # Preffered city
    location: Mapped[String] = mapped_column(String(), nullable=True)
    # Preffered car
    car_id: Mapped[int] = mapped_column(ForeignKey("cars.id"), nullable=True)

    user: Mapped[User] = relationship("User", back_populates="preferences")
    car: Mapped["Car"] = relationship("Car")
```

and schemas:
```py
class PreferencesRead(BaseModel):
    location: Optional[str] = None
    car_id: Optional[int] = None
    
class UserRead(schemas.BaseUser[uuid.UUID]):
    display_name: str
    avatar_url: Optional[str] = None
    birthday: Optional[datetime.date] = None
    preferences: PreferencesRead
```

**Reply by frankie567** [reply]
@Style77 You probably should enable `orm_mode` (or `from_attributes`) setting of Pydantic to allow the schema to read data from SQLAlchemy model:

* Pydantic 1.x: https://docs.pydantic.dev/1.10/usage/models/#orm-mode-aka-arbitrary-class-instances
* Pydantic 2.x: https://docs.pydantic.dev/2.2/api/config/#pydantic.config.ConfigDict.from_attributes

**Comment by Tur8008** [comment]
Hello! I need to change update_me method within get_user_router func -- that is still the case with phone confirmation. But I can't get how to adjust it properly as I dislike the idea to copy whole get_user_profile method to change only small submethod of it (update_me)
The example:
``` Python
 async def update_me(
        request: Request,
        user_update: user_update_schema,  # type: ignore
        user: models.UP = Depends(get_current_active_user),
        user_manager: BaseUserManager[models.UP, models.ID] = Depends(get_user_manager),
    ):
        try:
            user = await user_manager.update(
                user_update, user, safe=True, request=request
            )
            return user_schema.from_orm(user)
        except exceptions.InvalidPasswordException as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail={
                    "code": ErrorCode.UPDATE_USER_INVALID_PASSWORD,
                    "reason": e.reason,
                },
            )
        except exceptions.UserAlreadyExists:
            raise HTTPException(
                status.HTTP_400_BAD_REQUEST,
                detail=ErrorCode.UPDATE_USER_EMAIL_ALREADY_EXISTS,
            )
            # My code
        except MyCustomException as e:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="PHONE_IS_NOT_CONFIRMED"
``` 
I hope there is a nicer method to adjust this peace of code beside copying whole outer func. 

P.S. Thank you for the lib - it's really cool!

**Reply by frankie567** [reply]
In this case, no, you don't have the choice but to copy/paste the whole router. I would say it's not a really big deal, IMO. If we take a close look, it's actually just a call to `user_manager.update` and a bunch of `except`, there is not a lot of fancy/complicated logic there.

**Reply by Tur8008** [reply]
OK. I just wondered if there is a better way to implement overriding. Thanks for the answer!

**Reply by Tur8008** [reply]
```Python
app.include_router(
    get_users_router(get_user_manager, UserRead, UserUpdate, **authenticator** ,requires_verification=False),
    prefix="/users",
    tags=["users"],
)
```
I need to put Authenticator as a positional argument here. Where is it?

**Reply by Tur8008** [reply]
```Python
app.include_router(
    get_users_router(get_user_manager, UserRead, UserUpdate, fastapi_users.authenticator, requires_verification=False),
    prefix="/users",
    tags=["users"],
)
```

This way it works. Sorry for bothering!
